tests/bsim/bluetooth/host/att/long_read/main.c

```diff
@@ -0,0 +1,220 @@
+		/* The second PDU is one-less-than-full to test for off
+		 * by one errors.
+		 */
+		read_len -= 1;
+	}
+
+	/* If the ATT_MTU is too large, sending a one-less-than-full
+	 * response would exeed the max attribute length limit.
+	 */
+	__ASSERT(buf_len < (BT_ATT_MAX_ATTRIBUTE_LEN / 2),
+		 "The EATT buffer is too large for this test.");
+
+	/* Fill the read buffer with zero. */
+	memset(buf, 0, read_len);
+
+	/* Echo back the requested read size in the first two bytes of
+	 * each read.
+	 */
+	__ASSERT_NO_MSG(read_len >= 2);
+	sys_put_le16(read_len, buf);
+
+	return read_len;
+}
+
+static struct bt_gatt_attr attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(UUID_1),
+	BT_GATT_CHARACTERISTIC(UUID_2, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_mtu_validation_chrc, NULL, NULL),
+};
+
+static struct bt_gatt_service svc = {
+	.attrs = attrs,
+	.attr_count = ARRAY_SIZE(attrs),
+};
+
+static void find_the_chrc(struct bt_conn *conn, uint16_t *chrc_value_handle)
+{
+	uint16_t svc_handle;
+	uint16_t svc_end_handle;
+	uint16_t chrc_end_handle;
+
+	EZ(bt_testlib_gatt_discover_primary(&svc_handle, &svc_end_handle, conn, UUID_1, 1, 0xffff));
+
+	LOG_INF("svc_handle: %u, svc_end_handle: %u", svc_handle, svc_end_handle);
+
+	EZ(bt_testlib_gatt_discover_characteristic(chrc_value_handle, &chrc_end_handle, NULL, conn,
+						   UUID_2, (svc_handle + 1), svc_end_handle));
+
+	LOG_INF("chrc_value_handle: %u, chrc_end_handle: %u", *chrc_value_handle, chrc_end_handle);
+}
+
+static void bs_sync_all_log(char *log_msg)
+{
+	/* Everyone meets here. */
+	bt_testlib_bs_sync_all();
+
+	if (get_device_nbr() == 0) {
+		LOG_WRN("Sync point: %s", log_msg);
+	}
+
+	/* Everyone waits for d0 to finish logging. */
+	bt_testlib_bs_sync_all();
+}
+
+static inline void bt_enable_quiet(void)
+{
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_ERR);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_ERR);
+
+	EZ(bt_enable(NULL));
+
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_INF);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_INF);
+}
+
+static void test_long_read(enum bt_att_chan_opt bearer, uint16_t chrc_value_handle,
+			   struct bt_conn *conn)
+{
+	bool central = (get_device_nbr() == 0);
+
+	if (central) {
+		size_t read_count = 0;
+
+		NET_BUF_SIMPLE_DEFINE(attr_value, BT_ATT_MAX_ATTRIBUTE_LEN);
+
+		/* Perform the whole long read operation. */
+		EZ(btt_gatt_long_read(&attr_value, NULL, conn, bearer, chrc_value_handle, 0));
+
+		/* Parse the read attribute value to verify the
+		 * integrity of the transfer.
+		 *
+		 * Each response starts with the length of the whole
+		 * response and the rest is zero-padded.
+		 */
+		for (; attr_value.len; read_count++) {
+			uint16_t encoded_len;
+			uint16_t padding_size;
+
+			LOG_INF("Verifying read %u", read_count);
+
+			__ASSERT(attr_value.len >= sizeof(encoded_len),
+				 "Incomplete encoded length");
+			encoded_len = net_buf_simple_pull_le16(&attr_value);
+
+			padding_size = (encoded_len - 2);
```

Comment:
It's the size of the significant data in the read. I.e. the non-padding part.

Reply:
Not sure what "the significant data" refers to here. Can you elaborate?

Reply:
It's the 2 bytes making up the "encoded length" that is read out just above.

----------

I'm tempted to rewrite the test in simpler terms. It's becoming clear to me that the test code ended up too complex.

I will exercise the Zephyr Host API directly to perform long reads. This is better principled since I don't want the testlib to be an integral part of any test. It's meant to be used as just a helper to hide away the irrelevant bits.

I will try to come up with a test procedure that does not encode the read length. I think this may be the confusing expert coding part.

I may also add a scenario where the attribute length happens to align with the ATT_MTU. That corner case illustrates an easy-to-miss property of the long read procedure: Even if the response PDU is full, there's not necessarily more data to be read out. Rather, the client _does not know_ whether or not the end has been reached. The subsequent read can actually return a zero length result.

Reply:
> I may also add a scenario where the attribute length happens to align with the ATT_MTU. That corner case illustrates an easy-to-miss property of the long read procedure: Even if the response PDU is full, there's not necessarily more data to be read out. Rather, the client does not know whether or not the end has been reached. The subsequent read can actually return a zero length result.

I think that sounds good. For something like this (depending on how long time it takes to execute), it may even be worth to do it in a loop from reading 0 to 512 octets, and see that it works for any GATT attribute size.

Reply:
> It's the 2 bytes making up the "encoded length" that is read out just above.

Can we use `sizeof` so that it's clear what the 2 bytes represent?